docs/specs/user-export.md

```diff
@@ -0,0 +1,284 @@
+
+### The error response of Create an export
+
+> we have a global middleware of handling Admin API authentication,
+> that middleware returns 403 without a JSON body when authentication failed.
+
+|Description|Name|Reason|Info|
+|---|---|---|---|
+|When the rate limit exceeded|`TooManyRequest`|`RateLimited`|{"bucket_name": "UserExport"}|
+|When there is a running export|`TooManyRequest`|`MaximumConcurrentJobLimitExceeded`|-|
+|When the input fails the validation|`Invalid`|`ValidationFailed`|The info should contain the JSON schema validation output|
+|When the field names are not unique|`Invalid`|`UserExportNonUniqueFieldNames`|Output the full list of "field_names". Like `{"field_names": ["sub", "a", "b", "a"]}`|
+
+## Get the status of an export
+
+The endpoint is `GET /_api/admin/users/export/{ID}`
+
+- The endpoint requires Admin API JWT token to access.
+- This endpoint is added to Admin API server.
+- It is not part of the GraphQL API.
+- The result of a completed export lasts for 24h before it expires.
+- When the export is completed, the response body includes a freshly signed URL to the export file. The signed URL is valid for 60s.
+
+### The response body of Get the status of an export
+
+See [The response body](#the-response-body).
+
+### The error response of Get the status of an export
+
+> we have a global middleware of handling Admin API authentication,
+> that middleware returns 403 without a JSON body when authentication failed.
+
+|Description|Name|Reason|Info|
+|---|---|---|---|
+|When the given ID does not refer to an export|`NotFound`|`TaskNotFound`|-|
+
+## The response body
+
+The response body of a just created export looks like
+
+```
+{
+  "id": "some_opaque_string",
+  "status": "pending",
+  "created_at": "2024-01-01T00:00:00.000Z",
+  "request": {
+    "format": "ndjson"
+  }
+}
+```
+
+The response body of completed export looks like
+
+```
+{
+  "id": "some_opaque_string",
+  "status": "completed",
+  "created_at": "2024-01-01T00:00:00.000Z",
+  "completed_at": "2024-01-01T00:01:00.000Z",
+  "request": {
+    "format": "ndjson"
+  },
+  "download_url": "https://some-signed-url?with-a=signature"
+}
+```
+
+## The export file
+
+### The name of the export file
+
+The name of the export file is `{{ .AppID }}-{{ .TaskID }}-{{ .TaskCompletedAtInISO9601BasicFormat}}.{ndjson|csv}`.
+
+For example, given
+
+- The project is `myapp`.
+- The id of the task is `userexport_deadbeef`.
+- The completion time of the task is `2024-09-09T10:46:51.275Z`.
+- The requested `format` is `ndjson`.
+
+The name of the export file is `myapp-userexport_deadbeef-20240909104651Z.ndjson`.
+
+### The metadata of the export file
+
+- `Content-Disposition`: `attachment; filename=FILENAME`
+- `Content-Type`: `application/x-ndjson` for `ndjson`, `text/csv` for `csv`.
+
+Since the export file is accessed with a signed URL of a short validity, setting `Cache-Control` is not really helpful.
+
+### The housekeep of the export file
```

Comment:
What about on-perm deployment? Or there is a way on-perm deployment can disable this feature?

(So we might need another error code for disabled feature?)

Reply:
I intended to support cloud storage only. So on-perm deployment also requires a cloud storage bucket.

Reply:
do we need a new configure page for the setting up the cloud storage info of a project?

Reply:
In that case let's add a option to disable this feature, and the error code on the API when it's disabled?

So that for some on-perm deployment which we don't have cloud storage bucket, we can at least disable it instead of let it "hang there"...